[WIP] RPL border router sample application

jukkar opened the pull request:
This is a simple RPL based border router sample application. This can do routing between for example 802.15.4 and ethernet and act as a root node for RPL network. The router provides currently a very simple web UI for seeing the current nodes. Also new commands are added to net-shell to give better information about the RPL network.

jukkar committed:
samples: net: rpl: Simple RPL border router application

This is a very simple RPL border router sample application.
It provides HTTP(S) and net-shell interface for getting admin
information about neighbors and routes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,44 @@
+Building and Running
+********************
+
+Currently, the HTTP Server application is configured to listen at port 80.
```
s/Currently/By default/

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,44 @@
+To use QEMU for testing, follow the :ref:`networking_with_qemu` guide.
+
+This sample code supports both static and dynamic (DHCPv4) IP addresses that
+can be defined in the project configuration file:
```
suggest adding the name of the configuration file:

       in the project configuration file :file:`samples/net/rpl_border_router/prj.conf`:

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,44 @@
+	CONFIG_NET_APP_MY_IPV6_ADDR="2001:db8::1"
+	CONFIG_NET_APP_MY_IPV4_ADDR="192.0.2.1"
+
+Note that the IPv4 address is only used for connection to integrated web
```
connection to the integrated web

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,44 @@
+file and also in the :file:`samples/net/rpl_border_router/src/config.h` file.
+
+To use QEMU for testing, follow the :ref:`networking_with_qemu` guide.
+
```
shouldn't there be some commands here with the building instructions?

dbkinder reviewed (changes requested):
some suggested edits

jukkar commented:
Changes:
* Documentation fixes and enhancements.
* net-shell output fixes
* DAO-ACK support

nashif commented on samples/net/rpl_border_router/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+-   test:
+        build_only: true
+        platform_whitelist: frdm_k64f qemu_x86
```
if this sample is not intended for qemu, why are we whitelisting it here? why do we have a configuration for quark se but we are not building it? Also, the frdm board does not have 15.4.. I suggest removing the whitelist and adding:

depends_on: ieee802154

and make this more generic, we are going back to adding board specific configurations, something we need to stop doing.

jukkar replied:
I left the qemu_x86 there so it gets compile tested in qemu. Probably this is not needed if we are not running it in qemu.
I will add the depends_on line as suggested.

jukkar commented:
Updating the patches.

jukkar commented:
New version with RPL fixes.

jukkar commented:
Various bug fixes added, I will send these also in separate PR.

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,136 @@
+	shell> br repair
+	[rpl-br/shell] [INF] br_repair: Starting global repair...
+
+The **br repair** command will cause the RPL network to re-configure itself.
```
It's better to put the description of the command before the example rather than after on all of these examples.  For example, 

     The **br repair** command will cause the RPL network to re-configure itself.

     .. code-block:: console

        shell> br repair
        [rpl-br/shell] [INF] br_repair: Starting global repair...

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,136 @@
+
+	No parents found.
+
+The **net rpl** command first prints out static compile time configururation
```
move the description before the example (and fix the misspelling of "configuration"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,136 @@
+	[ 1] 0xa80065e0 1/0/1/0 0xa8007140 reachable      2920  00:12:4B:00:00:00:00:01 fe80::212:4b00:0:1
+	[ 2] 0xa8006660 1/0/1/0 0xa8007140 stale             0  00:12:4B:00:00:00:00:03 fe80::212:4b00:0:3
+
+The **net nbr** command prints information about currently found IPv6 neighbor
```
Move the description before the example.

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,136 @@
+	        neighbor  : 0xa8006660
+	        link addr : 00:12:4B:00:00:00:00:01
+
+The **nbr route** command prints information about currently found IPv6 routes.
```
Move the description before the example.

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,136 @@
+root node.
+
+If IEEE 802.15.4 support is enabled, then the **ieee15_4** command can be used
+to change the IEEE 802.15.4 network parameters like used channel or PAN id, if
```
s/like/such as/

"like" is used for comparison ("run like the wind"), and "such as" is use for examples

dbkinder reviewed (changes requested):
some recommended edits

jukkar commented:
New update:
* Fixed the documentation
* Various patches needed for debugging TCP

I will send all patches except the first border router patch, in a separate PR.

Current issues:
* no routing between network interfaces (yet)
* HTTP(S) causes hard faults in frdm_k64f board. These seem to be related to http_response_send_data() function support that were introduced in commit 64562e1af29cf46d9815aa861a8067f355425ad7. No crash errors are seen in QEMU when trying just the HTTP server part. Still debugging this issue.

jukkar commented:
All the other patches except the border router sample app are part of PR #1130

jukkar commented:
New version uploaded:

* routing still not implemented
* HTTP(S) now works ok.

jukkar committed:
net: ipv6: Add routing support between interfaces

Introduce CONFIG_NET_ROUTING option that allows the IP stack
to route IPv6 packets between multiple network interfaces.
No support for IPv4 routing is implemented by this commit.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ipv6: Add some extra debug when updating link address length

As this is very specialized info which is not normally needed,
do not print it by default.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version uploaded.

* IPv6 routing between network interfaces work
* Usage instructions updated
* Added example configuration files for frdm_k64f with cc2520 or mcr20a boards.

jukkar committed:
net: websocket: Initial support for server websocket

This commit creates a websocket library that can be used by
applications. The websocket library implements currently only
server role and it uses services provided by net-app API.
The library supports TLS if enabled in configuration file.

This also adds websocket calls to HTTP app server if websocket
connection is established.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket echo-server application

This is a http(s) server that supports also websocket.
It sends back any data sent to it over a websocket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version uploaded.

* integrates websocket support
* initial index.html + javascript to use the websocket added

jukkar commented:
Moved the routing patches to separate PR #1300 so that the set is easier to review.

jukkar committed:
net: websocket: Add console support

Add console driver that allows console session to be transferred
over a websocket connection.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket: Add console sample application

This sample application implements a web service that provides
zephyr console over websocket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
I was added as a reviewer to this PR, but I'm not familiar with RPL, so it mat take awhile before I'll be able to review this reasonably (after I do some study on RPL).

Nor (I think) I can't test this, and in this regard, I'd appreciate if @jukkar's test setup was described (so I know what I should target, I don't think I have needed/enough hardware for that).

Thanks.

jukkar committed:
net: rpl: Do not do neighbor discovery for RPL network

Doing neighbor discovery in RPL network is not necessary and
that can be disabled in RPL nodes. Unfortunately the border
router needs to have ND enabled as it has also non-RPL network
interfaces in use. So in this case, mark RPL node as always
reachable.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
> I was added as a reviewer to this PR, but I'm not familiar with RPL, so it mat take awhile before I'll be able to review this reasonably (after I do some study on RPL).

Ok, I removed you from review list.

pfalcon commented:
> Ok, I removed you from review list.

Thanks. No hints on this question? :

> I'd appreciate if @jukkar's test setup was described

jukkar committed:
net: rpl: Setup DAG prefix into network interface

This is needed in routing if we are acting as border router.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Uploaded new version that separates websocket from HTTP library. It also creates separate HTTP library that uses net-app library for connection management. Old HTTP library is still left as is but it may be deprecated in the future. I will split the patches to more logical pieces and send them separately for review.

nashif commented:
recheck

nashif commented:
why do we have changes to http and http legacy being added in this PR? Do we really need to introduce the _legacy.h variant here? I really dislike all those _legacy header files being added, this is just confusing, especially due the fact that http was just added a few releases back, so it is not really legacy...

andrewboie commented:
i don't think we should have all this legacy stuff lying around in the kernel for the LTS release, do we really want to have to support LTS obligations for two implementations of the same thing?

jukkar commented:
> why do we have changes to http and http legacy being added in this PR? Do we really need to introduce the _legacy.h variant here? I really dislike all those _legacy header files being added, this is just confusing, especially due the fact that http was just added a few releases back, so it is not really legacy...

Currently user never sees http_legacy.h file anyway, as the application just includes net/http.h which then selects proper include file according to Kconfig options. I made it like this so that we can deprecate the current http library more easily.

nashif commented:
> Currently user never sees http_legacy.h file anyway, as the application just includes net/http.h which then selects proper include file according to Kconfig options. I made it like this so that we can deprecate the current http library more easily.

You still have to configure this to use legacy in kconfig, so it is pretty much the same.

jukkar commented:
> You still have to configure this to use legacy in kconfig, so it is pretty much the same.

Currently developer gets to use the legacy one by default so no change in that respect to current behavior.

pfalcon commented on samples/net/http_app_server/src/main.c:
```diff
@@ -0,0 +1,397 @@
+	}
+
+	http_add_data(ctx, &pkt, HTTP_CRLF, strlen(HTTP_CRLF));
+	http_add_data(ctx, &pkt, payload, payload_len);
```
(Looking at what magic this app does that it requires https://github.com/zephyrproject-rtos/zephyr/pull/1330 and whether that can be addressed by https://github.com/zephyrproject-rtos/zephyr/pull/119 pretty easily.)

So, I won't be asking "what happens if payload is `static const payload[1000000]`" like I do in the mailing list RFC, let me just ask "what happens if payload_len is larger than free fragment space available, e.g. payload_len is 3K, but there're only 1.5K of free data buffers." You answer would be "But I allocated 8K of buffers above!" And that's cheating - the only way such paradigm would work is by overallocating buffers to back it.

I don't know where this "browser DOM" approach of "keep everything in memory" comes from. And the problem here is clearly in the API design. It's not natural for an embedded HTTP API to require a user to allocate a "network packet" and then keep adding data to it without limit. The natural API is `http_send_data(context *ctx, const char *buf, int len)`, and let that call decide what to do with that data: accumulate data into infinite buffer (typical "design pattern" for desktop/server apps nowadays), send even the smallest chunk of data right away (the opposite extreme), or buffer it per the natural size of buffer (which is single packet) to minimize copies, when it fills, send it, and repeat with the reminder of data - the approach I advocate in my RFC.

So, it looks like we have a case when "interesting" design of the HTTP API starting to affect design of the IP stack (like requiring #1330).

jukkar replied:
Please note that this is just a sample application that does things in certain way. Nothing prevents developer to send data in certain size chunks here. I just did not implement it here as #1330 was doing it for me automatically.

jukkar replied:
I sent updated HTTP API patches in #4243 
That PR differs slightly from the corresponding patches in this PR, for example the old HTTP API is deprecated, and client API support is added.

pfalcon replied:
> Please note that this is just a sample application that does things in certain way. Nothing prevents developer to send data in certain size chunks here. I just did not implement it here as #1330 was doing it for me automatically.

But #1330 is no longer an application patch, it's IP stack patch. If such patches go in, and such sample applications are in the mainline, it sends a strong message how things should be done in Zephyr, nobody simply will look for doing in another way. That's why it's important to have good generic APIs right away and resist desire to push workaround-style solutions down the stack.

pfalcon replied:
> I sent updated HTTP API patches in #4243

Thanks, will have a look.

jukkar commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+- Ethernet access for management purposes (optional).
+
+Note that there is no support for running RPL border router in QEMU, as the
+border router requires access to a real radio network technology like
```
We need more than one node in the network. The qemu emulation for 15.4 is one node only solution. This could be done in simulated way with emul8 but that is some future work.

jukkar commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+
+Note that the IPv4 address is only used for connection to the integrated web
+server that provides admin web page for management purposes. The web server
+can also be connected using IPv6 address. If you do not have a management
```
Management interface in this context refers to network interface that provides web support (==management operations) towards host computer. I obviously need to elaborate wording here.

dbkinder replied:
Yes, please do :)

jukkar commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+
+Note that the default project configuration file
+:file:`samples/net/rpl_border_router/prj.conf` does not currently provide
+a working system as there is no boards that would provide suitable network
```
Yes, that should work just fine.

jukkar commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+By default, the RPL border router application enables net shell support and
+provides some useful commands for debugging and viewing the network status.
+
+The **br repair** command will cause the RPL network to re-configure itself.
```
I wanted to keep the command short here thus "br" was used. Note that this command is only meant/used/available for this border router application so user probably understand what the command is for.

pfalcon replied:
> probably understand what the command is for.

Probably. I'd approach it from the other side though: "can it cause confusion or not?" Arguably it can. If short prefix is desired, I'd suggest "brt".

jukkar commented on samples/net/rpl_border_router/prj_frdm_k64f_cc2520.conf:
```diff
@@ -0,0 +1,106 @@
+CONFIG_NET_L2_IEEE802154_SHELL=y
+CONFIG_NET_L2_IEEE802154=y
+CONFIG_NET_L2_IEEE802154_FRAGMENT=y
+CONFIG_NET_APP_IEEE802154_CHANNEL=15
```
Just some channel used in testing, can be changed to default.

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: console
+
+	make BOARD=frdm_k64f
```
add:

     cd samples/net/http_server

dbkinder commented on samples/net/http_app_server/README.rst:
```diff
@@ -0,0 +1,298 @@
+
+.. code-block:: console
+
+        make BOARD=qemu_x86 run
```
add a ``cd`` to the right directory for this sample

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+
+Note that the default project configuration file
+:file:`samples/net/rpl_border_router/prj.conf` does not currently provide
+a working system as there is no boards that would provide suitable network
```
either "is no board" or "are no boards"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+Note that the default project configuration file
+:file:`samples/net/rpl_border_router/prj.conf` does not currently provide
+a working system as there is no boards that would provide suitable network
+interface support. The prj.conf file is provided only to compile test the
```
what does "compile test" mean?  you can compile but not run the sample? or did you mean to say "compile and test"?

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+	shell> ieee15_4 set_chan 15
+	Channel 15 set
+
+The border router sample application provides integrated HTTP(S) server.
```
provides an

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,169 @@
+
+The border router sample application provides integrated HTTP(S) server.
+Currently the admin support is very rudimentary but you can try it by connecting
+to http://192.0.2.1 or http://[2001:db8::1] using web browser.
```
using a

dbkinder commented on samples/net/ws_console/README.rst:
```diff
@@ -0,0 +1,40 @@
+
+The default make BOARD configuration for this sample is ``qemu_x86``.
+
+Connect to the console from you browser using these URLs http://[2001:db8::1]
```
from your browser

dbkinder commented on samples/net/ws_echo_server/README.rst:
```diff
@@ -0,0 +1,35 @@
+
+.. code-block:: console
+
+    $ cd $ZEPHYR_BASE/samples/net/websocket
```
should this be samples/net/ws_server ?

dbkinder commented on samples/net/ws_echo_server/README.rst:
```diff
@@ -0,0 +1,35 @@
+usage scenario is to run websocket-server application inside QEMU. This is
+described in :ref:`networking_with_qemu`.
+
+Build echo-server sample application like this:
```
should echo-server be websocket-server here?

dbkinder reviewed (changes requested):
suggested edits

jukkar committed:
net: tcp: Split sent packet into MSS size pieces

Instead of trying to send a packet that is larger than MSS,
split it into suitable pieces and queue the pieces individually.

Jira: ZEP-1998

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

galak commented:
What's the status for this, close it out?  Still aimed for 1.10?

jukkar committed:
tests: net: websocket: Add tests for websocket

This tests websocket by creating a websocket support http server
and sending data to it and verifying the returned data is the same.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Update:

* port to Cmake
* use latest HTTP and websocket APIs

nashif commented:
what is the status of this PR? Can it be closed? is it going to be updated and made ready for merge?

jukkar commented:
Ravi is currently working with these commits in #5035 so current plan is to close this PR after his set is updated (probably tomorrow).

rveerama1 commented:
I have separated RPL BR sample independent of patches and created another https://github.com/zephyrproject-rtos/zephyr/pull/5924. https://github.com/zephyrproject-rtos/zephyr/pull/5035 still contains all the patches. I will rebase and update https://github.com/zephyrproject-rtos/zephyr/pull/5035 once https://github.com/zephyrproject-rtos/zephyr/pull/5924 is merged.
